### DISCUSSION TITLE
Helix keymap rationale

### DISCUSSION THREAD

**Discussion by blitzerr** [helix-editor/helix#5442] [Q&A]
Hellix is awesome and a lot of awesomeness comes from reasonable defaults and shortcuts. I cannot remember all the shortcuts unless I understand the building blocks and then extend on those intuitively. But in some cases, I find myself unable to reason about shortcuts and I need your help.

Let me give an example. In insert mode, a very helpful shortcut is deleting characters / words in the forwards / backwards direction (There is something to be said about the lacking support of ctrl + [h, j, k, l] to navigate by one char in insert mode :). That's for another day). Usually I feel that there are these philosophies:
- Forwards is the normal direction and backwards is the alternate one. 
- If `<modifier> + <Key>` provides an action, then `<modifier> + <shift> + <Key>` provides the backwards version of it.
- If a `<Key>` provides an action, Alt + `<Key>` provides the same action, but backwards. Think `alt` as the `alternate`.

Now let's look at the delete shortcuts in the _insert_ mode:
| Action | Shortcut |
|---|----|
|Delete next word|Alt-d|
|Delete previous word | Ctrl-w|
|Delete next char|ctrl-d|
|Delete previous char|ctrl-h|
|Delete to end of line|Ctrl-k|
|Delete to start of line|Ctrl-u|

How do I reason about this ? Ctrl + d totally makes sense as this is also supported in most terminals and so our fingers do the right thing automatically. Now with ctrl + d deleting a character forward, I would expect ctrl + shift + d to delete a char backwards. Maybe alt + d deletes a word forward and alt + shift + d deletes the word backwards ? This is just one choice, there can be others and even better one. But the key is to have some underlying rationale between forwards / backwards operations and  their extensions to words, end of line and then another layer for N such items. And some building blocks (the most fundamental ones of course) transcend the mode boundaries like the keys h, j, k, l for navigation.

I am lacking the underlying intuition in the current choice. But because this is such a popular editor, I am sure people would have made the choices after great debates and weighing in options. So if you can tell me how you reason about this in your head, that would help me as well reason about it as well, and I can be more productive sooner. Thanks a lot !

**Comment by archseer** [comment]
Most of the insert mode bindings are so called "emacs bindings" that work similarly on a whole lot of other input fields. I personally don't recommend using any of these, they're mostly there for compatibility.

**Reply by blitzerr** [reply]
got it @archseer . So if one has to delete or navigate in the insert mode, the recommendation is to go to normal mode and then come back ? Is this also the recommendation for single character edits ?

**Reply by the-mikedavis** [reply]
Yep: insert mode is good for inserting text and using backspace/delete/`<C-w>` in insert mode is fine for small deletions at the cursor. More complicated deletions or changes should be done in normal mode. You may want to re-map the CapsLock key to Escape to make it very easy to switch between modes (that configuration is typically done at the operating system level)

**Reply by blitzerr** [reply]
Yes, I already have my caps lock set to escape. :)

**Reply by blitzerr** [reply]
@the-mikedavis  what is your recommendation for navigating by a single character. Say you pressed '(' and helix added a closing ')' and then placed your cursor between the two, but you don't have to provide any function arguments. 

Is the recommendation to press 
esc + l + a in this case ?

**Reply by archseer** [reply]
) will skip over the closing parenthesis

**Reply by blitzerr** [reply]
Got it. Thanks a lot @archseer
